File: Data Structure & Algorithm/Data Structure & Algorithm [Old]/Models/Sort/Direct Insertion Sort/Direct Insertion Sort.ts
// Direct Insertion Sort

/**
 * This is the core intuition behind insertion sort.
 * Given a collection of integers, you can sort the list by proceeding from the start of the list, and every time you encounter an element that is out of order, you can continuously swap places with previous elements until it is inserted in its correct relative location based on what you’ve processed thus far.
 *
 * Time Complexity: O(n^2)
 * Space Complexity: O(1)
 * It is a stable sorting algorithm.
 */

export function insertSort(nums: number[]) {
  for (let i = 1; i < nums.length; i++) {
    const current = nums[i];
    for (let j = i - 1; j >= 0; j--) {
      if (nums[j] > current) {
        nums[j + 1] = nums[j];
        nums[j] = current;
      }
    }
  }
}

export function insertSort1(nums: number[]): void {
  for (let i = 1; i < nums.length; i++) {
    const current = nums[i];
    let j = i - 1;
    while (j >= 0 && nums[j] > current) {
      nums[j + 1] = nums[j];
      j--;
    }
    nums[j + 1] = current;
  }
}

// test
const array = [5, 3, 1, 2, 4];
insertSort(array);
console.log(array);

File: Data Structure & Algorithm/Archive/Template/Sort/Direct Insertion Sort/Direct Insertion Sort.go
package main

import "fmt"

func insertSort(nums []int) {
	for i := 1; i < len(nums); i++ {
		current := nums[i]
		j := i - 1
		for j >= 0 && nums[j] >= current {
			nums[j+1] = nums[j]
			j--
		}
		nums[j+1] = current
	}
}

// test
func main() {
	array := []int{10, 1, 3, 2, 9, 1, 5, 6}
	insertSort(array)
	fmt.Println(array)
}
